gwcs/spectroscopy.py

```diff
@@ -192,3 +196,223 @@ def input_units(self):
             return {'wavelength': 1 / self.groove_density.unit,
                     'alpha_in': u.Unit(1),
                     'beta_in': u.Unit(1)}
+
+
+class Snell3D(Model):
+    """
+    Snell model in 3D form.
+
+    Inputs are direction cosines.
+
+    Parameters
+    ----------
+    n : float
```

Comment:
reverting back to my original comment. A model can define the types of inputs and outputs as floats, and then combine models, but we don't normally do that for parameters though. That is, I don't think the existing formalism allows compound models where a model is fed into a parameter. So shouldn't n be a model input if we want to use the other models that compute the index of refraction?

Reply:
Yah, that's how I wrote it. Though when I started I thought I will write it with `n` being an input. I forgot in the two weeks I stopped work on it. 😞